spec/fixtures/custom_plugins/kong/plugins/tcp-trace-exporter/handler.lua

```diff
@@ -0,0 +1,104 @@
+local cjson = require "cjson"
+local str = require "resty.string"
+
+local ngx = ngx
+local kong = kong
+local table = table
+local insert = table.insert
+local to_hex = str.to_hex
+
+local _M = {
+  PRIORITY = 1001,
+}
+
+local tracer_name = "tcp-trace-exporter"
+
+function _M:rewrite(config)
+  if not config.custom_spans then
+    return
+  end
+
+  local tracer = kong.tracing(tracer_name)
+
+  local span = tracer.start_span("rewrite", {
+    parent = kong.tracing.active_span(),
+  })
+  tracer.set_active_span(span)
+end
+
+
+function _M:access(config)
+  local tracer = kong.tracing(tracer_name)
+
+  local span
+  if config.custom_spans then
+    span = tracer.start_span("access")
+    tracer.set_active_span(span)
+  end
+  
+  kong.db.routes:page()
+
+  if span then
+    span:finish()
+  end
+end
+
+
+local function push_data(premature, data, config)
+  if premature then
+    return
+  end
+
+  local tcpsock = ngx.socket.tcp()
+  tcpsock:settimeout(1000)
+  local ok, err = tcpsock:connect(config.host, config.port)
+  if not ok then
+    kong.log.err("connect err: ".. err)
+    return
+  end
+  local _, err = tcpsock:send(data .. "\n")
+  if err then
+    kong.log.err(err)
+  end
+  tcpsock:close()
+end
+
+function _M:log(config)
+  local tracer = kong.tracing(tracer_name)
+  local span = tracer.active_span()
+
+  if span then
+    kong.log.debug("Exit span name: ", span.name)
+    span:finish()
+  end
+
+  kong.log.debug("Total spans: ", ngx.ctx.KONG_SPANS and #ngx.ctx.KONG_SPANS)
+
+  local spans = {}
+  local process_span = function (span)
+    local s = table.clone(span)
+    s.tracer = nil
+    s.parent = nil
+    s.trace_id = to_hex(s.trace_id)
+    s.parent_id = s.parent_id and to_hex(s.parent_id)
+    s.span_id = to_hex(s.span_id)
+    insert(spans, s)
```

Comment:
Hmm. I know this is just a test fixture plugin, but it seems like plugins that make use of tracing will need to implement similar functionality. I wonder if it makes any sense to wrap some of this up in a `span:serialize()` method (or similar) so that callers don't need to be aware of the need to 1) copy the table and 2) `nil`-ify specific, non-serializeable fields like `tracer` and `parent`.

Reply:
Good point, I think this makes sense and is more friendly for callers, but I think it's expensive to copy the whole table, I'd like to add more convenience to the pdk until we have a best practice, e.g. the coming open telemetry plugin.